### ISSUE TITLE
USING FTPSERVER with SDFat.h lib causes: error:'File' is already declared in this scope using fs::File

### ISSUE THREAD

**Issue by axfatec** [nailbuster/esp8266FTPServer#18]
Hello. I'm a new user of esp8266, and sice few months ago I started to develop a data logger with ESP-12E module.
All worked since I decided to add the FTP feature to my project (where I use library #include "SdFat.h")
Now I added your library <ESP8266FtpServer.h> and when I compile the scketch I obtain this error:

sp8266/FS.h:136:11: error: 'File' is already declared in this scope

using fs::File;
^
I found here me-no-dev/ESPAsyncWebServer#112 , an old discussion on similar issue, but I was now able to implement to your libary and to solve the issue.
Could you help me ?
thanks Alessandro

**Comment by tobozo** [comment]
There was a [breaking change](https://github.com/esp8266/Arduino/issues/1524) on the SD library due to keywords collisions with SPIFFS. Both were using the `File` keyword differently.

Now old SD keywords need to be prefixed i.e. `fs::` or`sd::`

Depending on the complexity of your other project you have two choices:

1) update your SD library to the oldest version using the Arduino library manager, you can still rollback to the newer SD library version later.

2) update your other sketch so it uses the new namespaced keywords, it's just a matter of adding `sd::`, `fs::` or `dir::` to the existing keywords, compilation errors will help you to find where it needs to be done.

**Comment by axfatec** [comment]
I found a working solution of SD.h lib reworked by : http://mcutry.tistory.com/47
that seems to re-addresses to SD the calls to SPIFFS.
Unfortunately this lib has not been merged to ufficial one. it's a pitty.
Could anyone with right experience to do this improvment? ..   and furthermore, it should be better in this case to update the SDFat library.

**Comment by tobozo** [comment]
Well it's using the namespaced version too, and the problem has already been partially adressed (you still need to edit your sketch though) by using a define just before including `SD.h`: `#define FS_NO_GLOBALS`

**Comment by axfatec** [comment]
Better ! thanks.
so in my case (I'm not so expert)   should have I to change my calls 
 if(!SD.begin(chipSelect)) 
to like this?
 if(!SD::SD.begin(chipSelect))   

and similarly to:
myFile=SD.open("DATA.txt", FILE_WRITE);    in  myFile=SD::SD.open("DATA.txt", FILE_WRITE);    ??

**Comment by tobozo** [comment]
not exactly, SD syntax is unchanged so you can keep this:

`myFile = SD.open( .... );`

however when using SPIFFS, you'll have to use this:

`fs::myFile = SPIFFS.open( ..... );`

**Comment by axfatec** [comment]
Thank you a lot..!
I will check this new lib with "ESP8266FtpServer.h",  then I will report any news.

**Comment by axfatec** [comment]
Do you know if there is a porting to SDFat.h library in order to use it together with ESP8266FtpServer.h solving the SD card  access  instead of SPIFFS access ?
This should be a better solution because SDFat can mange long file name (other than improved functions respect to SD.h)
Thanks

**Comment by tobozo** [comment]
I only use SDFat.h with Atmega chips, essentially because it has a small memory footprint and can fit in a bootloader.
And I always remove the library after use as it often conflicts with other non-atmega libraries (and this is probably what's happening to you).

Now you don't have to modify SD.h; this tutorial does not use a specific SD.h library other than the one you can get from your library manager: http://mcutry.tistory.com/47

The article hint you at modifying the actual esp8266FTPServer library so it calls SD methods instead of SPIFFS methods, and this implies some small changes.

If you comment out every #include "FS.h" and implement the changes from the article (marked in yellow), it should work as described, provided you selected the correct version of the library in the Arduino library manager.

**Comment by axfatec** [comment]
Actually, the SD.h modified by the article, with ESP8266FTPServer , works correctly.
What I would do, (but I found hard) is to modify in the same way the SDFat.h library, but due to different metods adopted in this one comprare to Sd.h ,  it doesn't work.  Most probably I did some mistakes.
Could you share your SDFat.h library so I can test it with ESP8266FTPserver  ?
If it could works, it should be a very good achievement.
Thanks
Alex

**Comment by tobozo** [comment]
> Could you share your SDFat.h library so I can test it with ESP8266FTPserver ?

It's the [official SDFat.h](https://github.com/greiman/SdFat), I haven't tried it with ESP8266 

The official updated SD.h from the Arduino Library manager is better for your needs, and it does not default to 8.3 filenames, why wouldn't you use it? It does implement the changes you're talking about after all.

I just diffed the modified library from the article, and there's only one change in one file: File.cpp
```
#> diff File.cpp File.cpp.bak
17,19d16
< using namespace sd;
< using sd::File;
< 
```

If you look at ~/Arduino/libraries/SD/File.cpp from the Arduino library folder (SD Built-in by Arduino SparkFun) you'll see those changes are there too, only they're at the end of the file.

So basically you can delete the custom SD directory from your ~/Arduino/libraries folder, go to your Arduino Library manager, search "SD", select the version `1.1.1` of "SD Built-in by Arduino SparkFun", update the library, and still get the same results as in the article.

**Comment by tobozo** [comment]
btw from the [network page](https://github.com/nailbuster/esp8266FTPServer/network) it seems like somebody already did this a [year ago](https://github.com/nailbuster/esp8266FTPServer/compare/master...apullin:sd-card)

**Comment by axfatec** [comment]
But I tried and it did't work. So, something is still missing with joined use of SDFAt and ESP8266FTPServer.

Anyway, I will try your previuous suggestion.  Did you furthermore told that new SD lib support long file name??  It should be great.
